text/0000-grand-testing-unification.md

```diff
@@ -0,0 +1,551 @@
+# Grand Testing Unification -- RFC
+
+- Start Date: 2016-02-05
+- RFC PR: [emberjs/rfcs#119](https://github.com/emberjs/rfcs/pull/119)
+- Ember Issue: N/A
+
+# Summary
+
+The testing story in Ember today is better than it ever has been. It is now possible to test individual component/template combos, register your own mock components/services/etc, build complex acceptance tests, and most anything else you would like.
+
+Unfortunately, there is a massive disparity between different types of tests.  In acceptance tests, you use well designed global helpers to deal with async related interactions; whereas in integration and unit tests you are forced to manually deal with this asynchrony. The goal of this RFC is to unify the concepts amongst the various types of test (acceptance, integration, and unit) and provide a single common structure to tests.
+
+# Motivation
+
+Usage of component integration style tests is becoming more and more common, but these tests still include manual event delegation (`this.$('.foo').click()` for example), and assumes most (if not all) interactions are synchronous.  This is a major issue due to the fact that the vast majority of interactions will actually be asynchronous. There have been a few recent additions to `ember-test-helpers` that have made dealing with asynchrony better, but forcing users to manually manage all async is a recipe for disaster.
+
+Acceptance tests allow users to handle asynchrony with ease, but they rely on global helpers that automatically wrap a single global promise which makes testing of interleaved asynchronous things more difficult. There are a number of limitations in acceptance tests as compared to integration tests (cannot mock and/or stub services, cannot look up services to setup test context, etc).
+
+We need a single unified way to teach and understand testing in Ember that leverages all the things we learned with the original acceptance testing helpers that were introduced in Ember 1.0.0 and improved in 1.2.0.  Instead of inventing our own syntax for dealing with the async (`andThen`) we should use new language features such as `async` / `await`.
+
+# Detailed design
+
+The goal of this RFC is to introduce new constructs, one for each type of test, that all share the same basic structure and helper system. This new system will be implemented in the `ember-test-helpers` library so that it can iterate faster while supporting multiple Ember versions independently.
+
+This proposal will implement a few core concepts to address these issues:
+
+- Usage of `async`/`await` semantics
```

Comment:
related tracking issue for async/await by default https://github.com/ember-cli/ember-cli/issues/3529